packages/realm/src/IterableWeakSet.ts

```diff
@@ -0,0 +1,128 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { binding } from "./internal";
+
+/**
+ * Implements an iterable `WeakSet`.
+ * @internal
+ */
+export class IterableWeakSet<T extends object> implements Set<T> {
```

Comment:
This is a bit of a weird type. I don't think it is correct to say it `implements Set<T>`, even if it has the right signatures. In addition to the O(1) vs O(N) issue, it also has the wrong semantics because `add(x)` always adds even if there is already a value equal to `x` in the set (due to wrapping in a new WeakRef). Consider:

```js
const mySet = new IterableWeakSet();
const someObj = {};
mySet.add(someObj);
mySet.add(someObj); // should be a no-op but isn't!
assert(mySet.size == 1); // Boom!
```